src/rdysetup.c

```diff
@@ -202,88 +210,160 @@ static PetscErrorCode FindSalinityCondition(RDy rdy, const char *name, PetscInt
   PetscFunctionReturn(0);
 }
 
-// initializes mesh region/surface data
-static PetscErrorCode InitRegionsAndSurfaces(RDy rdy) {
+// initializes mesh region data
+static PetscErrorCode InitRegions(RDy rdy) {
   PetscFunctionBegin;
 
   // Count and fetch regions.
   PetscInt c_start, c_end;  // starting and ending cell points
   PetscCall(DMPlexGetHeightStratum(rdy->dm, 0, &c_start, &c_end));
   DMLabel label;
   PetscCall(DMGetLabel(rdy->dm, "Cell Sets", &label));
-  for (PetscInt region_id = 0; region_id <= MAX_REGION_ID; ++region_id) {
-    IS cell_is;
-    PetscCall(DMLabelGetStratumIS(label, region_id, &cell_is));
-    if (cell_is) ++rdy->num_regions;
-    PetscCall(ISDestroy(&cell_is));
-  }
-  PetscCall(RDyAlloc(PetscInt, rdy->num_regions, &rdy->region_ids));
-  PetscCall(RDyAlloc(RDyRegion, rdy->num_regions, &rdy->regions));
-  PetscInt r = 0;
-  for (PetscInt region_id = 0; region_id <= MAX_REGION_ID; ++region_id) {
-    IS cell_is;  // cell index space
-    PetscCall(DMLabelGetStratumIS(label, region_id, &cell_is));
-    if (cell_is) {
-      RDyRegion *region  = &rdy->regions[r];
-      rdy->region_ids[r] = region_id;
-      ++r;
-
-      PetscInt num_cells;
-      PetscCall(ISGetLocalSize(cell_is, &num_cells));
-      if (num_cells > 0) {
-        rdy->region_ids[rdy->num_regions] = region_id;
-        region->num_cells                 = num_cells;
-        PetscCall(RDyAlloc(PetscInt, region->num_cells, &region->cell_ids));
-      }
-      const PetscInt *cell_ids;
-      PetscCall(ISGetIndices(cell_is, &cell_ids));
-      for (PetscInt i = 0; i < num_cells; ++i) {
-        region->cell_ids[i] = cell_ids[i] - c_start;
-      }
-      PetscCall(ISRestoreIndices(cell_is, &cell_ids));
+  if (label) {  // found regions (cell sets) in the grid
+    for (PetscInt region_id = 0; region_id <= MAX_REGION_ID; ++region_id) {
+      IS cell_is;
+      PetscCall(DMLabelGetStratumIS(label, region_id, &cell_is));
+      if (cell_is) ++rdy->num_regions;
       PetscCall(ISDestroy(&cell_is));
     }
+    PetscCall(RDyAlloc(PetscInt, rdy->num_regions, &rdy->region_ids));
+    PetscCall(RDyAlloc(RDyRegion, rdy->num_regions, &rdy->regions));
+    PetscInt r = 0;
+    for (PetscInt region_id = 0; region_id <= MAX_REGION_ID; ++region_id) {
+      IS cell_is;  // cell index space
+      PetscCall(DMLabelGetStratumIS(label, region_id, &cell_is));
+      if (cell_is) {
+        RDyRegion *region  = &rdy->regions[r];
+        rdy->region_ids[r] = region_id;
+        ++r;
+
+        PetscInt num_cells;
+        PetscCall(ISGetLocalSize(cell_is, &num_cells));
+        if (num_cells > 0) {
+          rdy->region_ids[rdy->num_regions] = region_id;
```

Comment:
Should this be `rdy->region_ids[r-1] = region_id`? (`r-1` because `r` was incremented on [238L](https://github.com/RDycore/RDycore/pull/24/files#diff-75d4508176a97a2c171e581ffa0d15d8d268b032c241fe78fef679b7d55652aaR238)).

Reply:
On second thought, this line should be deleted as you have already set the value on 237L.

Reply:
Good find. Actually, I think we can get rid of that line, because line 237 already does what this one intends.